service.go

```diff
@@ -255,6 +274,96 @@ func (svc *Service) createResponse(initialEvent *nostr.Event, content interface{
+		} else {
+			startOfWeek = now.AddDate(0, 0, -int(weekday)+1)
+		}
+		return time.Date(startOfWeek.Year(), startOfWeek.Month(), startOfWeek.Day(), 0, 0, 0, 0, startOfWeek.Location())
+	case "monthly":
+		return time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
+	case "yearly":
+		return time.Date(now.Year(), time.January, 1, 0, 0, 0, 0, now.Location())
+	default: //"never"
+		return createdAt
+	}
+}
+
+func GetEndOfBudget(budget_type string, createdAt time.Time) time.Time {
+	start := GetStartOfBudget(budget_type, createdAt)
+	switch budget_type {
+	case "daily":
+		return start.AddDate(0, 0, 1)
+	case "weekly":
+		return start.AddDate(0, 0, 7)
+	case "monthly":
+		return start.AddDate(0, 1, 0)
+	case "yearly":
+		return start.AddDate(1, 0, 0)
+	default: //"never"
+		return start.AddDate(9999, 0, 0)
+	}
+}
+
+func (svc *Service) hasPermission(app *App, event *nostr.Event, requestMethod string, paymentRequest *decodepay.Bolt11) (result bool, code string, message string) {
+	// find all permissions for the app
+	appPermissions := []AppPermission{}
+	findPermissionsResult := svc.db.Find(&appPermissions, &AppPermission{
+		AppId:     app.ID,
+	})
+	if findPermissionsResult.RowsAffected == 0 {
+		// No permissions created for this app. It can do anything
+		svc.Logger.Info("No permissions found for app", app.ID)
+		return true, "", ""
+	}
+
+	appPermission := AppPermission{}
+	findPermissionResult := findPermissionsResult.Limit(1).Find(&appPermission, &AppPermission{
+		RequestMethod: requestMethod,
+	})
+	if findPermissionResult.RowsAffected == 0 {
+		// No permission for this request method
+		return false, NIP_47_ERROR_RESTRICTED, fmt.Sprintf("This app does not have permission to request %s", requestMethod)
+	}
+	ExpiresAt := appPermission.ExpiresAt
+	if !ExpiresAt.IsZero() && ExpiresAt.Before(time.Now()) {
+		svc.Logger.Info("This pubkey is expired")
+		return false, NIP_47_ERROR_EXPIRED, "This app has expired"
+	}
+	MaxAmountPerTransaction := appPermission.MaxAmountPerTransaction
+	if MaxAmountPerTransaction != 0 {
+		if paymentRequest.MSatoshi/1000 > int64(MaxAmountPerTransaction) {
+			svc.Logger.Info("trying to send more than max amount per transaction in budget")
+			return false, NIP_47_ERROR_INSUFFICIENT_BALANCE, "Payment amount is greater than budget allows"
+		}
+	}
+	
+	maxAmount := appPermission.MaxAmount
+	if maxAmount != 0 {
+		budgetUsage := svc.GetBudgetUsage(&appPermission)
+		
+		if budgetUsage +paymentRequest.MSatoshi/1000 > int64(maxAmount) {
+			return false, NIP_47_ERROR_QUOTA_EXCEEDED, "Insufficient budget remaining to make payment"
+		}
+	}
+	return true, "", ""
+}
+
+func (svc *Service) GetBudgetUsage(appPermission *AppPermission) int64 {
+	var result SumResult
+	svc.db.Table("payments").Select("SUM(amount) as sum").Where("app_id = ? AND preimage IS NOT NULL AND created_at > ?", appPermission.AppId, GetStartOfBudget(appPermission.BudgetRenewal, appPermission.App.CreatedAt)).Scan(&result)
```

Comment:
@bumi @kiwiidb what are your thoughts on the preimage check here (originally added by Bumi)? is it so that failed payments are ignored? couldn't it cause more payments than the budget to get through if they are sent simultaneously?

Reply:
tbh I feel like ignoring failed payments is more important than being really strict about the budget.

Reply:
What about if we check the payment has a preimage or the payment was created at least one minute ago? or do you think this is not necessary?